proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+|``wrong-do-bind``                              |default|   E   |   E  |  E  |  E   |          |        |       |        |        |     |   |  |      |        |        |  |  |  |                                  |
++-----------------------------------------------+-------+-------+------+-----+------+----------+--------+-------+--------+--------+-----+---+--+------+--------+--------+--+--+--+----------------------------------+
+
+Effect and Interactions
+-----------------------
+
+By leaning on language editions, we achieve the following:
+
+* Users can declare, at the top of their file or in their cabal file,
+  that they are working in the stable subset of Haskell.
+
+* By clearly describing the stable subset of Haskell, this proposal
+  allows us to uphold the stability principles `articulated
+  elsewhere <https://github.com/ghc-proposals/ghc-proposals/blob/wip/general-rules/principles.rst#3ghc-stability-principles>`_.
+
+* If we ever feel the need to evolve the meaning of an extension, we
+  can do so, by keying the meaning of the extension to the language
+  edition. This allows forward evolution without sacrificing stability.
+  (There may be some implementation duplication to achieve this. There
+  is a cost to stability, but one I think is worth paying.)
+
+* We have a mechanism for introducing new warnings. Over the years,
+  we've been skittish about adding new warnings on by default, lest
+  they cause trouble for older programs. With language editions, we
+  can tell whether a program is old or new, enabling new warnings only
+  in a newer edition (if we so choose).
+
+* There is no need for a language edition simply to be a collection
+  of language extensions and warnings. We might imagine features
+  that just live in an edition, without a separate flag controlling
+  it. (Indeed, this happens in this original proposal a bit, by
+  splitting up different aspects of ``-XImpredicativeTypes``.) If
+  language editions prove popular, they open up new ways of evolving
+  GHC without the fuss of language extensions.
+
+* Coding tools (e.g. editor integrations, HLS) can choose to target,
+  say, just the stable subset of Haskell. While support for all of
+  the language is preferred, of course, having editions helps provide
+  guidance to tool authors for where to focus their efforts.
+
+* The ``Stable`` edition is somewhat restrictive, with warnings that
+  cannot be disabled and extensions that cannot be turned on. Yet
+  there is always an escape hatch: use ``Experimental`` instead.
+  The motivation here is that we want ``Stable`` to mean exactly that.
+  If a ``Stable`` module were allowed to twiddle settings that violate
+  the stability goals, then the user might unintentionally make their
+  program unstable. By making ``Stable`` restrictive, we force the
+  user to be aware of the stability impact of their settings.
```

Comment:
Ah -- but the risks do not apply only to that user. That's the problem with stability. If I declare that my package `p` is stable (by writing `Stable2024` in the cabal file, say) and then you write a package that depends on `p`, you are taking on the risk I signed up for. That's problematic. This worry underpins the design here around restrictiveness in various areas.

Reply:
I like Tom's gloss on this a lot.

I think this is touching on a key point, similar to the "transitive-or-not" discussion in https://github.com/ghc-proposals/ghc-proposals/pull/617. _Enforcing_ that a package is only using `Stable` (and presumably _it's_ dependencies?) is a big deal, and requires something more like Safe Haskell (including potentially writing information into interface files). But if you want a UX that is similar to "using a whole different compiler", then that's what you need.

In contrast, just using these editions to _locally_ control what stuff is available is much simpler.

Reply:
Additionally... perhaps there is room for both mechanisms. If we were to adapt the Safe Haskell machinery to track stability, we could imagine "stability inference" that could do a pretty good job: if you use `Stable2024` everywhere and don't use any unstable extensions anywhere, then we can stamp a module as stable.

Reply:
This sounds reasonable, though I expect this scheme to be about as widely used as `Safe` itself.

Reply:
Perhaps the lack of demand for `Safe` indicates a lack of demand for this kind of mechanism in general :shrug:

Reply:
Thanks, @tomjaguarpaw, for elucidating our difference so crisply. I completely agree with your characterization.

The challenge here is that I keep hearing complaints (in general) about churn in our ecosystem. I believe that unless we enforce stability of packages that claim it, we will not eliminate the problem. What's challenging here is the lack of voices of actual users. I've posted on Discourse looking for some, but that didn't seem to work. I'm sure the sheer size of this proposal is part of the reason.... but I don't see another way to get the idea across that won't (rightly) elicit "but this needs more detail" from the committee.

In direct response to the Safe-like infrastructure: that could work in theory, but I don't think it will in practice -- part of why I have the stance I do here. If a promise of stability is enforced, we don't need any trust mechanism.

Reply:
I wonder whether there's a hidden assumption here that the stable fragment of the language is too small to meet the needs of most users. I don't share that assumption. But maybe others are operating with it? If so, that's a problem too big for this proposal to solve.

Reply:
I think a tricky thing about discussions around ecosystem churn is that it's difficult to quantify the cause of the churn. I've done a few upgrades of my workplaces large-ish codebase to new version of GHC (8.10->9.0, (then quickly) 9.0->9.2, 9.2->9.4, 9.4->9.8).

I feel like in discussions a lot of the blame is placed on GHC/core-libraries. Sometimes this is fair. For instance the `SimplifiedSubsumption` changes were unexpectedly difficult to handle. Yet, often users (like me) only upgrade packages when upgrading GHC. So, 1.5 years of breaking changes from the entire ecosystem have to be dealt with at once. Often as well, maintainers (understandably) tend to bundle releases of necessary dependency bumps with breaking changes, so you get even more churn. Stackage somewhat makes this worse as well, since it encourages an all-or-nothing approach. I think issues in this area are much harder to quantify and less easily actionable than breaking changes in GHC or core libraries, so are often given less attention.

In my view quite a strong Safe like infrastructure would not avoid ecosystem churn completely (though it would help somewhat). This is because for any realistic application you have many dependencies, and even if they all guarantee that they will only use stable parts of GHC the language, it's quite likely that some dependency will have some sort of breaking change in its interface. Or there might be breaking changes in C FFI dependencies, or upstream protocols, etc.

I think very strong language stability guarantees can be quite helpful for other contexts such as for educational code, or examples in books. But for production code, you normally want to use near to the latest versions of libraries, so churn is somewhat unavoidable, since changes to GHC the language are just one source of breaking changes.

I think https://github.com/ghc-proposals/ghc-proposals/pull/636#issuecomment-1968868773 makes an excellent point in terms of the costs of enforcing this stuff.